sdk/storage/azure-storage-blob-nio/src/main/java/com/azure/storage/blob/nio/NioBlobInputStream.java

```diff
@@ -0,0 +1,201 @@
+
+    private final BlobInputStream blobInputStream;
+
+    NioBlobInputStream(BlobInputStream blobInputStream) {
+        this.blobInputStream = blobInputStream;
+    }
+
+    /**
+     * Returns an estimate of the number of bytes that can be read (or skipped over) from this input stream without
+     * blocking by the next invocation of a method for this input stream. The next invocation might be the same thread
+     * or another thread. A single read or skip of this many bytes will not block, but may read or skip fewer bytes.
+     *
+     * @return An <code>int</code> which represents an estimate of the number of bytes that can be read (or skipped
+     * over) from this input stream without blocking, or 0 when it reaches the end of the input stream.
+     */
+    @Override
+    public synchronized int available() {
+        return this.blobInputStream.available();
+    }
+
+    /**
+     * Closes this input stream and releases any system resources associated with the stream.
+     */
+    @Override
+    public synchronized void close() {
+        this.blobInputStream.close();
+    }
+
+    /**
+     * Marks the current position in this input stream. A subsequent call to the reset method repositions this stream at
+     * the last marked position so that subsequent reads re-read the same bytes.
+     *
+     * @param readlimit An <code>int</code> which represents the maximum limit of bytes that can be read before the mark
+     * position becomes invalid.
+     */
+    @Override
+    public synchronized void mark(final int readlimit) {
+        this.blobInputStream.mark(readlimit);
+    }
+
+    /**
+     * Tests if this input stream supports the mark and reset methods.
+     *
+     * @return Returns {@code true}
+     */
+    @Override
+    public boolean markSupported() {
+        return this.blobInputStream.markSupported();
+    }
+
+    /**
+     * Reads the next byte of data from the input stream. The value byte is returned as an int in the range 0 to 255. If
+     * no byte is available because the end of the stream has been reached, the value -1 is returned. This method blocks
+     * until input data is available, the end of the stream is detected, or an exception is thrown.
+     *
+     * @return An <code>int</code> which represents the total number of bytes read into the buffer, or -1 if there is no
+     * more data because the end of the stream has been reached.
+     * @throws RuntimeException when no available bytes to read.
+     * @throws IOException If an I/O error occurs.
+     */
+    @Override
+    public int read() throws IOException {
+        try {
+            return this.blobInputStream.read();
+        } catch (RuntimeException e) {
+            if (e.getMessage().equals(Constants.STREAM_CLOSED)) {
+                throw LoggingUtility.logError(logger, new IOException(e));
+            }
+            throw LoggingUtility.logError(logger, e);
```

Comment:
Shoundn't this already happen inside BlobInputStream ? If not would it be possible to push it down there instead of creating new type ?

Reply:
BlobInputStream throws RuntimeExceptions, and the contract specifies that these methods are supposed to throw an IOException. I suspect that BlobInputStream was changed to throw RuntimeExceptions to work better with Reactor, but in an environment that isn't touched by async stuff, I think it's best to conform to the actual api conctract. 

I can look into skipping the double logging if you don't like that. It'll probably require a checkstyle suppression since CI doesn't usually let you use a raw throw statement without somehow passing it through a logger.

Reply:
I'm not sure if BlobInputStream throwin RuntimeException is/was intentional. As you said the InputStream contract forces caller to handle IOException, so reactive code we have that's calling it should be handling/convering IOExceptions (otherwise compiler wouldn't be happy) - if that's the case I think we could safely make underlying stream confirming to the contract instead (e.g. throw IOException on certain response codes).

If we keep error translation here then I think we have more to cover . I.e. underlying stream can throw `new RuntimeException(UNEXPECTED_STREAM_READ_ERROR)` . Then, if the list of messages indicating IOException grows then we'll have a challage of maintaining parity between this class and underlying infra (which would be another argument to push this responsibility to underlying stream).

Reply:
I think changing the kinds of exceptions we throw is a breaking change, even if the customer theoretically should be handling it already, no?

Reply:
Since we're talking about `RuntimeException` vs `IOException` I think it all boils down to how did we document `@throws` in existing streams.

For example [here](https://github.com/Azure/azure-sdk-for-java/blob/895bf6f96ca5926e6071f5055e95101e2cb133dc/sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/StorageInputStream.java#L228-L229) or [here](https://github.com/Azure/azure-sdk-for-java/blob/895bf6f96ca5926e6071f5055e95101e2cb133dc/sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/StorageInputStream.java#L193) the described contract is quite open. Moreover I'd say that if we throw `RuntimeExceptions` there on something that's "I/O related" then that should be bug (since we don't conform to our own documentation).